File: cours-js/scopes.js
// en JS let permet d'inscire une variable dans un scope local
// tandis que var permet d'inscrire une variable dans une scope semi-local

for (let i = 0; i < 10; i++) {
    console.log(i);
}
// si la variable i a était définie avec le mot clé let.
// i n'est accessible que si elle a été définie dans la même scope ou un scope extérieur
console.log(i);

for (var i = 0; i < 10; i++) {
    console.log(i);
}

// si la variable i a été définie avec le mot clé var, elle est accessible même si elle a été définie dans un scope intérieur
console.log(i);

function myFunc1() {
    var myVar1 = 123;
}

// même si la variable myVar1 est déclarée avec le mot clé var, comme elle est déclarée à l'intérieur d'une fonction, elle n'est pas accessible depuis un scope extérieur
// ReferenceError: myVar1 is not defined
//console.log(myVar1);

 //_____________________________________//  LES CLOSURE //___________________________________//

 function multNFactory(n) {
    return function multN(x) {
        return x * n;
    }
 }

 mult3 = multNFactory(3);

 for (let i = 0; i < 3; i++) {
        console.log(mult3(i));
 }

 mult54 = multNFactory(54);

 for (let i = 0; i < 3; i++) {
    console.log(mult54(i));
}

 //_____________________________________//  LES CLOSURE //___________________________________//

File: cours python/scopes.py
#________________________________________________________________________________________#

                    # scope (étendue)= portée des variables #

#________________________________________________________________________________________#

my_var = 123

def my_func1():
    print(my_var)
    print(my_var2)

my_var2 = 42

# la fonction voit les variables définies à l'extérieur au préalable ou à postériori

my_func1()

#________________________________________________________________________________________#

def my_funct1():
    my_var3 = 3.14

# my_func2()

# il n'est pas possible d'accéder à une variable définie à l'intérieur d'une fonction, que celle-ci est était executée ou non
# principe du verre tinté (comme une limousine)
# NameError: name 'my_var3' is not defined. Did you mean: 'my_var'?

# print(my_var3)

#________________________________________________________________________________________#

my_var4 = 'foo'

def my_func3(my_var4):

    # le paramètre my_var4 masque la variable définie à l'extérieur de la fonction 

    print(my_var4)

# cet appel affiche 'bar'

my_func3('bar')

#________________________________________________________________________________________#

my_var4 = 'foo'

def my_func4():
    # la variable de my_var4 fait de l'ombre à la variable définie à l'extérieur de la fonction 
    my_var4 = 'baz'
    print(my_var4)

my_func4()

# la variable de my_var4 définie à l'extérieur de la fonction reste inchangée

print(my_var4)

#________________________________________________________________________________________#

def my_func5(myvar5):
    my_var5 = 'foo'

my_var6 = 123

# les variables de types int, float, bool ou str sont passés par valeur
# c-à-d que la fonction pourra accéder à la variable originale définie à l'extérieur

my_func5(my_var6)
print(my_var6)

#________________________________________________________________________________________#

def my_func6(my_var7: list):
    my_var7.append('foo')

my_var8 = [123, 42, 3.14]

# les variables de types list, dict, tuple ou objet sont passées par référence
# c-à-d que la fonction pourra accéder à la variable originale définie à l'extérieur

my_func6(my_var8)
print(my_var8)

#________________________________________________________________________________________#